flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/BatchCompiledPlanTestBase.java

```diff
@@ -0,0 +1,262 @@
+        program.getSetupTemporalFunctionTestSteps().forEach(s -> s.apply(tEnv));
+
+        final CompiledPlan compiledPlan;
+        if (program.runSteps.get(0).getKind() == TestKind.STATEMENT_SET) {
+            final StatementSetTestStep statementSetTestStep = program.getRunStatementSetTestStep();
+            compiledPlan = statementSetTestStep.compiledPlan(tEnv);
+        } else {
+            final SqlTestStep sqlTestStep = program.getRunSqlTestStep();
+            compiledPlan = tEnv.compilePlanSql(sqlTestStep.sql);
+        }
+
+        compiledPlan.writeToFile(path);
+    }
+
+    @ParameterizedTest
+    @MethodSource("createSpecs")
+    @Order(1)
+    void loadAndRunCompiledPlan(TableTestProgram program, ExecNodeMetadata metadata)
+            throws Exception {
+        final EnvironmentSettings settings = EnvironmentSettings.inBatchMode();
+        settings.getConfiguration().set(StateBackendOptions.STATE_BACKEND, "rocksdb");
+        final TableEnvironment tEnv = TableEnvironment.create(settings);
+        tEnv.getConfig()
+                .set(
+                        TableConfigOptions.PLAN_RESTORE_CATALOG_OBJECTS,
+                        TableConfigOptions.CatalogPlanRestore.IDENTIFIER);
+
+        program.getSetupConfigOptionTestSteps().forEach(s -> s.apply(tEnv));
+
+        for (SourceTestStep sourceTestStep : program.getSetupSourceTestSteps()) {
+            final Collection<Row> data = sourceTestStep.dataBeforeRestore;
+            final String id = TestValuesTableFactory.registerData(data);
+            final Map<String, String> options = new HashMap<>();
+            options.put("connector", "values");
+            options.put("data-id", id);
+            options.put("runtime-source", "NewSource");
+            options.put("terminating", "true");
+            options.put("bounded", "true");
+            sourceTestStep.apply(tEnv, options);
+        }
+
+        for (SinkTestStep sinkTestStep : program.getSetupSinkTestSteps()) {
+            final Map<String, String> options = new HashMap<>();
+            options.put("connector", "values");
+            options.put("disable-lookup", "true");
+            options.put("sink-insert-only", "false");
+            sinkTestStep.apply(tEnv, options);
+        }
+
+        program.getSetupFunctionTestSteps().forEach(s -> s.apply(tEnv));
+        program.getSetupTemporalFunctionTestSteps().forEach(s -> s.apply(tEnv));
+
+        final CompiledPlan compiledPlan =
+                tEnv.loadPlan(PlanReference.fromFile(getPlanPath(program, metadata)));
+
+        compiledPlan.execute().await();
+        for (SinkTestStep sinkTestStep : program.getSetupSinkTestSteps()) {
+            List<String> expectedResults = getExpectedResults(sinkTestStep, sinkTestStep.name);
+            assertThat(expectedResults)
+                    .containsExactlyInAnyOrder(
+                            sinkTestStep
+                                    .getExpectedBeforeRestoreAsStrings()
```

Comment:
What is this line good for? Ideally, in batch mode we merge before and after restore rows together and materialize changelogs. Then we can test for equality.

Reply:
There is `getExpectedAsStrings` already.

Reply:
> Ideally, in batch mode we merge before and after restore rows together and materialize changelogs. Then we can test for equality.

Ok, there are two parts to this; let's separate them.

First, we can test with either the before data or both before and after combined.  I picked the former since it didn't seem to matter whether or not we used all the of the data.

Second, materializing the changelogs is harder than it seems.  Presently, we are not providing a clear indication what the upsert keys are.  Also, changelogs may be in upsert or retract formats.  And they can be expressed as strings.  

Given that, when it came time to materialize changelogs, I just cheated and added the materialization as a separate array.  

See this example from my draft PR: 
https://github.com/apache/flink/pull/25063/files#diff-ada28191d8115d27cbcbda5549f2b12399c7021bc578041607bd5dc23d3774e3L57-R63

For the first point, I am indifferent to whether we test with more or less of the data.  On the second point, I think that just writing down the materialization in the Test Program manually is going to be the "quickest" way to do things.